kvdb/interface.go

```diff
@@ -93,6 +93,42 @@ type RwCursor = walletdb.ReadWriteCursor
 // writes. When only reads are necessary, consider using a RTx instead.
 type RwTx = walletdb.ReadWriteTx
 
+// ExtendedRTx is an extension to walletdb.ReadTx to allow prefetching of keys.
+type ExtendedRTx interface {
+	RTx
+
+	// RootBucket returns the "root bucket" which is pseudo bucket used
+	// when prefetching (keys from) top level buckets.
+	RootBucket() RBucket
+}
+
+// ExtendedRBucket is an extension to walletdb.ReadBucket to allow prefetching
+// of all values inside buckets.
+type ExtendedRBucket interface {
+	RBucket
+
+	// Prefetch will attempt to prefetch all values under a path.
+	Prefetch(paths ...[]string)
```

Comment:
I think this could use a better explanation of what `Prefetch` really does from the caller point of view. It loads the given values/ranges in memory and holds them until they are requested via a `Get` in the same transaction? And probably also good to describe how this interacts with all the other bucket operations (`Delete`, `CreateBucket`, etc) and whether the data remains consistent.

An extension of the generic kvdb-level tests would be a useful safe-guard too here and helps with documenting the behavior as well.

I am also wondering how to implement prefetching for Postgres. The bucket interface as proposed here gives a lot of freedom to call `Prefetch` where ever one wants. I think the implementation can get quite complicated (and risky) if the prefetched keys need to be kept consistent.

Isn't there a more minimalistic alternative that works just as well for performance? One idea that I had is to convert `Prefetch` to something like `GetMulti` that directly returns all the buckets/data ranges that were fetched in an optimized way. Then it is clear that the data is just that and doesn't keep on living in a background cache that needs to be kept consistent.

Reply:
`GetMulti` could work, but in case where we also want to spare the roundtrips that come from just nesting into buckets how would that work semantically? In this case we prefetch based on the payment hash, fetching several bucket keys so whenever we nest we won't go to the DB in the same transaction.

Reply:
Just a `GetMulti` that gets all keys in a bucket should be quite simple, but I don't know how much of the bottleneck will still be left because indeed the bucket also needs to be reached first. Could be worth testing.

For including the down traversal in the same transaction too, `GetMulti` might take as an input a tree of keys perhaps. In that case, can't you do the exact same thing that you do currently, but just hidden? So in the implementation of `GetMulti` for `etcd`, you'd internally call `prefetch` first followed by a (recursive) series of `get` calls to populate the result tree that you mention?

For postgres, `GetMulti(tree)` in a single tx would be an interesting challenge. Either magic with subqueries, or a change to the db structure to make it possible to retrieve deeply nested keys in one step (using the sha256 ids perhaps).

For errors, I'd think that just returning an error from `GetMulti` can be enough because currently we probably also don't do anything special when a condition happens that isn't supposed to happen ever? Missing keys is probably happy flow, but I think that can be signaled in the result tree.

If something like this can work, I think it is worth looking into. The `Prefetch` approach may work for client-side transactions, but I think it gets hard when connected to database backends that manage the transaction server-side.

Reply:
I'm wondering if we actually need the `Prefetch` for Postgres? I recall having stable performance on Postgres where seemingly roundtrips did not bring any slowdowns over time?

Reply:
How about adding the `Prefetch` and `GetMulti` as well? Where `GetMulti` would simply work on the current bucket so we'd at least avoid the roundtrip overhead coming from the cursor?

Reply:
Postgres doesn't need it for stability as far as I know, but reducing round-trips should surely increase performance. If the current gain for `etcd` can also be achieved with a slightly different interface that is compatible with server-side transactions, I think it is a win.

Has the root cause of the etcd slowdown been found in the mean time? Maybe the prefetching fixes the problem on some machines, but not on others. Or there are other unknown factors that could cause it to return again in nodes with a different life than the benchmark nodes.

Reply:
If just `GetAll` is sufficient for `postgres`, I'd think that it should also be sufficient for `etcd`? The round-trip itself likely is the most costly part and I'd think that round-trip counts should be comparable across backends.

Also it would probably make the delta of `stm.go` smaller and save time in adding sufficient test coverage for all the ways in which you can use `Prefetch`.

Reply:
Seems this PR is moving to the final stages, but shouldn't the `Prefetch` API be compatible with database backends that support transactions server-side? @Roasbeef

Reply:
I think the idea is that we have `Prefetch` as optional requirement (see `kvdb/interface.go`) and we could extend (in a separate PR) with `GetAll` as suggested. The reason is that `Prefetch` can make it possible to fetch a payment in one roundtrip while `GetAll` can optimize on `ForEach` calls.

Reply:
If for optimal performance it is required to get a payment in one round trip (makes sense), how can a separate PR that adds `GetAll` deliver that? It still won't be a single round trip. Implementing `Prefetch` as it is currently proposed isn't feasible for server-side transaction-based systems in my opinion.